docs/investigations/gateway-architecture/gateway-architecture.md

```diff
@@ -0,0 +1,116 @@
+# Gateway architecture
+
+## Overview
+
+We discussed three different approaches to exposing our API to users.
+
+1. Maintaining one handwritten GraphQL API on gateway, that would call all other microservices (it could use for example gRPC)
+2. Stitching multiple GraphQL schemas written for internal services into a single one and proxying the traffic to those internal GraphQL servers
+3. Having separate HTTP endpoints proxying traffic to specific internal servers (exposing many separate GraphQL APIs)
+
+## Possible solutions
+
+### 1. Handwritten GraphQL Gateway
+
+First option is to use single GraphQL server on gateway with manually written schema and resolvers, that would delegate the operations to internal services. We could use for example gRPC to communicate with them (a PoC of this can be found [here](https://github.com/kyma-incubator/compass/pull/21/)).
+
+Pros
+
+- full GraphQL support (introspection + subscriptions)
+- avoiding Node.js (better performance)
+- single endpoint
+- facade over internal APIs
+
+Cons
+
+- need to maintain almost identical protobuf schema to relay requests to internal services
+- each resolver sends separate call to proxied API
```

Comment:
it's good to note that if we used gRPC, active connection will be maintained, so the cost of remote call is not that high.